HIP/hip-792.md

```diff
@@ -0,0 +1,290 @@
+
+(2) Smart contract execution occurs exclusively "on-chain",
+and therefore a smart contract may not provide its authorization of a transaction
+by any means other than its own execution when invoked within a transaction.
+In other words smart contract authorization must necessarily occur while a transaction is "in flight",
+and therefore cannot occur client-side before being submitted to the network.
+
+However the latter scenario with smart contract IDs
+within "complex" keys **should** be supported, as noted in
+[Hedera's documentation on Keys](https://docs.hedera.com/hedera/sdks-and-apis/hedera-api/basic-types/key):
+
+> Note that when a Key is a smart contract ID,
+> it doesn't mean the contract with that ID will actually create a cryptographic signature.
+> It only means that when the contract calls a precompiled contract,
+> the resulting "child transaction" will be authorized to perform any action controlled by the Key.
+
+Therefore this HIP proposes a means, via the `hederaAccountSystem` system contract,
+to verify authorization of a transaction that will work in the above scenarios.
+
+> NOTE:
+>
+> Hedera Token Service already exposes a system contract `hederaTokenService`,
+> and its function already performs the same authorization as proposed
+> for `hederaAccountService.isAuthorizedCurrentTransaction()`
+>
+> This works as system contracts, including `hederaTokenService`,
+> are passed in a [`TxnAwareEvmSigsVerifier`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/sources/TxnAwareEvmSigsVerifier.java#L60),
+> which has a [`TransactionContext`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/sources/TxnAwareEvmSigsVerifier.java#L70)
+> that contains the necessary information in [`JKey`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/legacy/core/jproto/JKey.java#L37).
+> <!-- TODO verify this about TransactionContext in system contracts -->
+
+## User stories
+
+<!-- Provide a list of “user stories” to express how this feature, functionality, improvement, or tool will be used by the end user. Template for user story: “As (user persona), I want (to perform this action) so that (I can accomplish this goal).” -->
+
+(1)
+As a smart contract developer,
+I want to code a smart contract in solidity which is able to identify whether a transaction is authorized when sent from an account with a complex key which contains a smart contract ID,
+so that I can implement advanced use cases including:
+
+- implement flexible and customisable authorisation rules
+- implement advanced multisig use cases without the need to split across multiple transactions
+- implement atomic multisig use cases
+- implement batch processing of multiple transactions
+
```

Comment:
The batch processing use case is clear to me but the others less so.  Are there more concrete examples of why this function is necessary to enable those use cases as opposed to simply failing if the signature is insufficient?

Reply:
The batch scenario is actually an interesting one where let's say there are 3 accounts, 

AccA --> `ThresholdKey([ecDsaKeyA, customSc], 1)`
AccB --> `ThresholdKey([ecDsaKeyB, customSc], 1)`
AccC --> `ThresholdKey([ecDsaKeyC, customSc], 1)`

They could individually sign 3 separate transactions with their cryptographic keys.

Let's say that `customSc` can perform that same transaction when passed in an array of account addresses.
Since all these accounts require 1-of-2, and `customSc`'s smart contract ID is among their keys, those indiidual transactions can be performed in a single transaction as a batch.

(To be clear, this can also be accomplished with `isAuthorized` and `isAuthorizedRaw` from HIP-632 too, with clients taking a different approach)